gensim/models/keyedvectors.py

```diff
@@ -558,6 +564,134 @@ def most_similar(self, positive=None, negative=None, topn=10, restrict_vocab=Non
+    case_insensitive=True, dummy4unknown=False, topk=5):
+        """Compute performance of the model on an analogy test set using a set based method.
+
+        The accuracy is reported (printed to log and returned as a score) for each section separately,
+        plus there's one aggregate summary at the end. It implements set based 3CosAvg method.
+        It has been implemented in a similar way with the original evaluate_word_analogies method
+        which is for the 3CosAdd OR 3CosMul methods. According to 3CosAvg method we compute
+        the average of the difference between every given pair of words and then we add it
+        in the source word of the last pair trying to predict the target word of the last pair.
+        This method corresponds to the `compute-accuracy` script of the original C word2vec.
+        See also `3CosAvg <https://www.aclweb.org/anthology/C/C16/C16-1332.pdf>`_. For more
+        details see  on GitHub #2321 <https://github.com/RaRe-Technologies/gensim/pull/2321>`_.
+
+        Parameters
+        ----------
+        analogies : str
+            Path to file, where lines are tuples of words, split into sections by ": SECTION NAME" lines.
+            See `gensim/test/test_data/questions-words.txt` as example.
+        restrict_vocab : int, optional
+            Ignore all tuples containing a word not in the first `restrict_vocab` words.
+            This may be meaningful if you've sorted the model vocabulary by descending frequency (which is standard
+            in modern word embedding models).
+        case_insensitive : bool, optional
+            If True - convert all words to their uppercase form before evaluating the performance.
+            Useful to handle case-mismatch between training tokens and words in the test set.
+            In case of multiple case variants of a single word, the vector for the first occurrence
+            (also the most frequent if vocabulary is sorted) is taken.
+        dummy4unknown : bool, optional
+            If True - produce zero accuracies for tuples with out-of-vocabulary words.
+            Otherwise, these tuples are skipped entirely and not used in the evaluation.
+        topk: int, optional
+            Number of the top similar words we can consider for a successful prediction.
+
+        Returns
+        -------
+        (float, list of dict of (str, (str, str, ...., str))
+            Overall evaluation score and full lists of correct and incorrect predictions divided by sections.
+
+        """
+        ok_vocab = [(w, self.vocab[w]) for w in self.index2word[:restrict_vocab]]
+        ok_vocab = {w.upper(): v for w, v in reversed(ok_vocab)} if case_insensitive else dict(ok_vocab)
+        oov = 0
+        logger.info("Evaluating word analogies for top %i words in the model on %s", restrict_vocab, analogies)
+        sections, section = [], None
+        tuples = 0
+        for line_no, line in enumerate(utils.smart_open(analogies)):
```

Comment:
This method is doing too much. You have a parser for the analogies file integrated with logic for calculating analogies, performing case normalization, and other things.

This is bad for several reasons:

1. The code is not reusable
2. The code is difficult to test (BTW, where are the unit tests for this functionality? e.g. parsing the analogies file.
3. The function is long, nested and difficult to keep in reader's memory in one piece
4. Poor separation of concerns

My recommendation: refactor this. Start by splitting out the parser logic (e.g. splitting the file into sections).

Reply:
@mpenkov the logic of the implementation is the same as for the quadruples analogy questions. Could you please clarify what do you mean by "splitting the file into sections" and explain what is the difference with the case of the quadruples (not set based computations). Thanks.